core/runtime/runtime_context.hpp

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include "log/logger.hpp"
+#include "outcome/outcome.hpp"
+#include "primitives/common.hpp"
+#include "runtime/types.hpp"
+#include "storage/trie/types.hpp"
+
+namespace kagome::storage::trie {
+  class TrieBatch;
+}
+
+namespace kagome::blockchain {
+  class BlockHeaderRepository;
+}
+
+namespace kagome::storage::changes_trie {
+  class ChangesTracker;
+}
+
+namespace kagome::runtime {
+  class ModuleFactory;
+  class ModuleInstance;
+  class ModuleRepository;
+
+  class RuntimeContext {
+   public:
+    // should be created from runtime contex factory
+    RuntimeContext() = delete;
+    RuntimeContext(const RuntimeContext &) = delete;
+    RuntimeContext &operator=(const RuntimeContext &) = delete;
+
+    RuntimeContext(RuntimeContext &&) = default;
+
+    // constructor for tests
+    static RuntimeContext create_TEST(
+        std::shared_ptr<ModuleInstance> module_instance) {
+      return RuntimeContext{module_instance};
+    }
+
+    struct ContextParams {
+      ContextParams() = delete;
+
+      MemoryLimits memory_limits;
+    };
+
+    const std::shared_ptr<ModuleInstance> module_instance;
+
+   private:
+    friend class RuntimeContextFactoryImpl;
+    friend class RuntimeContextFactory;
+    RuntimeContext(std::shared_ptr<ModuleInstance> module_instance);
+  };
+
+  class RuntimeContextFactory {
+   public:
+    using ContextParams = RuntimeContext::ContextParams;
+
+    virtual ~RuntimeContextFactory() = default;
+
+    static outcome::result<RuntimeContext> fromCode(
+        const runtime::ModuleFactory &module_factory,
+        common::BufferView code_zstd,
+        ContextParams params = {});
```

Comment:
Does `ContextParams{}` conflict with `ContextParams() = delete`?

Reply:
It actually doesn't, but I will remove ContextParams() = delete because it is no longer needed.